[P/D]Provide bucket algorithm rate limiter  for proxy_server

frankie-ys opened the pull request:
# Essential Elements of an Effective PR Description Checklist
 * [x]  The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
 * [ ]  The test plan, such as providing test command.
 * [ ]  The test results, such as pasting the results comparison before and after, or e2e results
 * [ ]  (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
 
 ## Purpose
I found that run vllm with 1P1D with disaggregation example, the proxy server don't have rate limiter. When request's concurrency bigger than 20, the prefill or decode insatence will hangs or crash. After add rate limiter to proxy server , it works smoothly. And solve this issue 's problem  which because of high concurrency request. [https://github.com/vllm-project/vllm/issues/11247](url)
What's more, you can contact me via email if you have any questions. 
 ## Test Plan
 No need to add new test.
 
 ## Test Result
 
 
 ## (Optional) Documentation Update

frankie-ys committed:
[BugFix] Provide bucket algorithm rate limiter and queue for 1P1D proxy to handle concurrent requests and prevent the prefill or decode service crashed or hangs (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[PR Fix] prevent deadLock (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[PR Fix] fix ruff (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[PR Fix] fix ruff (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[PR Fix] fix ruff (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[PR Fix] fix ruff (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[PR Fix] fix ruff (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[PR Fix] fix ruff (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys commented:
I found that an error of OOM happened on buildkite/fastcheck/pr. Can this check get a manual trigger?
<img width="2499" height="1428" alt="image" src="https://github.com/user-attachments/assets/a287d84e-e883-4e2d-8848-2aa00931087a" />

frankie-ys committed:
[PR Fix] fix ruff (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys commented:
> Retrying

Unfortunatly，it happened again. How can I solve this problem?  
<img width="2541" height="1446" alt="image" src="https://github.com/user-attachments/assets/59aaddf6-928e-4a97-95b6-62f42e334c02" />

frankie-ys committed:
Merge branch 'vllm-project:main' into main

frankie-ys committed:
[PR Fix] fix ruff (#22575)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

DarkLight1337 commented on benchmarks/disagg_benchmarks/disagg_prefill_proxy_server.py:
```diff
@@ -1,63 +1,196 @@
+REQUEST_QUEUE_SIZE = 500  # Maximum number of requests in the queue
+RATE_LIMIT = 40  # Maximum requests per second (rate limiting)
+# Prefill service endpoint
+PRE_SERVICE_URL = "http://localhost:8100/v1/completions"
```
```suggestion
PREFILL_SERVICE_URL = "http://localhost:8100/v1/completions"
```

I would rename this to be consistent with the decode one

frankie-ys committed:
Update benchmarks/disagg_benchmarks/disagg_prefill_proxy_server.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>
Signed-off-by: frankie <wangyongsheng686@gmail.com>

frankie-ys committed:
Update benchmarks/disagg_benchmarks/disagg_prefill_proxy_server.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>
Signed-off-by: frankie <wangyongsheng686@gmail.com>

frankie-ys commented on benchmarks/disagg_benchmarks/disagg_prefill_proxy_server.py:
```diff
@@ -1,63 +1,196 @@
+
+                # Refill tokens if more than 1 second has passed
+                if elapsed > 1.0:
+                    self.tokens = self.rate_limit
```
Thanks,  I will fix it.

frankie-ys committed:
[review Fix]
Modify the variables and abstract the request queue into a separate file (#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[review Fix]  abstract the rate limiter into a separate file (#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[review Fix]  abstract the rate limiter into a separate file (#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] ruff fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] ruff fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] ruff fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

DarkLight1337 commented on benchmarks/disagg_benchmarks/disagg_prefill_proxy_server.py:
```diff
@@ -1,63 +1,127 @@
 async def forward_request(url, data):
+    """Forward request to backend service with rate limiting and error handling"""
+    # Apply rate limiting before making request
+    await rate_limiter.acquire()
```
I think it would be more idiomatic to make this a context manager

frankie-ys replied:
Thanks! I have made the rate limiter as context manager.

frankie-ys committed:
[Fix] add rate limiter as context manager(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] add rate limiter as context manager(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

DarkLight1337 reviewed (commented):
LGTM now but I'll have @KuntaiDu do the final review

frankie-ys committed:
[Fix] fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

KuntaiDu commented on benchmarks/disagg_benchmarks/disagg_prefill_proxy_server.py:
```diff
@@ -1,63 +1,130 @@
+logger = logging.getLogger(__name__)
+
+# Timeout for backend service requests (seconds)
+AIOHTTP_TIMEOUT = aiohttp.ClientTimeout(total=300)
+# Maximum concurrent requests to backend services
+MAX_CONCURRENT_REQUESTS = 100
+REQUEST_QUEUE_SIZE = 500  # Maximum number of requests in the queue
+RATE_LIMIT = 40  # Maximum requests per second (rate limiting)
```
It would be nice if we can move these to cli args.

frankie-ys replied:
Yeah, I have moved the variable to the cli args.

KuntaiDu reviewed (commented):
Other part LGTM.

frankie-ys committed:
[Fix] ruff fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
Merge branch 'main' into main

frankie-ys committed:
[Fix] ruff fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

frankie-ys committed:
[Fix] fix(#22643)

Signed-off-by: frankie-ys <yongshengwang@cmbchina.com>

KuntaiDu reviewed (approved):
The code looks much much cleaner. LGTM!
But since currently there are people fixing CI, so I will postpone the merge till the CI is green.

DarkLight1337 commented:
Can you merge from main branch?

KuntaiDu committed:
Merge branch 'main' into main

KuntaiDu commented:
Sure! Thanks for letting me know!
